scipy/cluster/hierarchy.py

```diff
@@ -1052,29 +1027,33 @@ def linkage(y, method='single', metric='euclidean', optimal_ordering=False):
             raise ValueError("Method '{}' requires the distance metric "
                              "to be Euclidean".format(method))
         if y.shape[0] == y.shape[1] and np.allclose(np.diag(y), 0):
-            if np.all(y >= 0) and np.allclose(y, y.T):
+            if xp.all(y >= 0) and xp.allclose(y, y.T):
                 _warning('The symmetric non-negative hollow observation '
                          'matrix looks suspiciously like an uncondensed '
                          'distance matrix')
         y = distance.pdist(y, metric)
+        y = xp.asarray(y)
     else:
         raise ValueError("`y` must be 1 or 2 dimensional.")
 
-    if not np.all(np.isfinite(y)):
+    if not xp.all(xp.isfinite(y)):
         raise ValueError("The condensed distance matrix must contain only "
                          "finite values.")
 
     n = int(distance.num_obs_y(y))
     method_code = _LINKAGE_METHODS[method]
 
+    y = np.asarray(y)
```

Comment:
Worth adding a comment is this convert to numpy -> do computations -> convert back is the idiom for interfacing with custom extensions (is this the reason it's done here?)

Reply:
Yep this is what I wanted to say in the description of the PR. When going to Cython you need to convert to NumPy. And once you are done you can come back to your backend.

I wrote a context manager if we want to use that instead, maybe that could help with the task 😃